doc/usecases/core/UC_keyset_lookup_cascading.md

```diff
@@ -0,0 +1,35 @@
+# Use Case: Cascading Lookup in `KeySet`
+
+## Summary
+
+- **Title:** Cascading Lookup in `KeySet`
+- **Scope:** `libelektra-core`
+- **Level:** Developer Goal
+- **Actors:** Core, Caller
+- **Brief:** Caller looks up `Key` with cascading name in existing `KeySet`
+
+## Scenarios
+
+- **Precondition:**
+  - [`KeySet` has been created](UC_keyset_create.md)
+  - (Only Alternative) Specification for desired `Key` exists and contains links
+- **Main success scenario:**
+  - Caller asks Core to look up `Key` by cascading name in `KeySet`
+  - Core checks for specification matching desired `Key`
+  - Core does not find specification
+  - Core tries equivalent name in all appropriate namespaces in correct order
+  - If a matching `Key` is found, Core returns the index
+  - Otherwise, Core returns a value indicating "Not found".
+    No valid index is returned, because `Key`s with cascading names cannot be inserted.
+- **Alternative scenario:**
+  - Caller asks Core to look up `Key` by cascading name in `KeySet`
+  - Core checks for specification matching desired `Key`
+  - Core finds specification
+  - Core tries links from specification and equivalent names in appropriate namespaces in correct order
```

Comment:
It's unclear whether handling the specification links should even be the core's responsibility. It's not ideal, e.g. adding new kinds of links requires modifying the core. IMO this is also more part of a separate group of use cases concerned with specifications in general.

If we remove this alternative scenario here, I would make this cascading lookup case simply an alternative scenario of the direct lookup.

Reply:
I was thinking about lookup plugins a lot and currently don't think it is worth the effort. I think the current lookup can be used as fixed integrated part of the core.

Reply:
I wouldn't actually create separate lookup plugins. IMO the cascading lookup makes sense. But the `meta:/fallback` et al. links are a bit out of place in the core IMO. It is basically the only place where metadata has an actual effect on a core function (which kinda makes it no longer **meta**data).

What I was thinking about was more like having a function pointer somewhere that points to a function in the spec plugin (i.e. the currently active implementation of the spec hook) when the `KeySet` comes from `kdbGet`. In other `KeySet`s it doesn't point anywhere, but I don't think the link stuff has any use for other keysets.

Some options for where to put the function pointer (just OTTOMH, no thought put into options):
- A pointer in `KeySet`; called when a cascading lookup finds no result; arguments are the `KeySet *` and the key we're looking for
- In `meta:/callback/lookup` instead of the current links; called when the key is found in a cascading lookup; arguments are the `KeySet *` and the key we're looking for
- Add new namespace `KEY_NS_LOOKUP`; has highest priority in cascading lookup; always contains function pointer; called when the key is found in a cascading lookup; arguments are the `KeySet *` and the key we're looking for

Like I said, I just find it a bit weird that the core knows about this stuff, when it makes almost no sense, if you don't know it's used for configuration data.

Reply:
> the current solution that the lookup key has an undocumented function pointer as value is questionable.

Where is that the case?

> Can you please open a decision about that.

Not right now no. There are enough open decisions and we're making zero progress. This is not high priority stuff IMO.